packages/react-data-grid/src/common/editors/EditorContainer.tsx

```diff
@@ -1,274 +1,246 @@
+import TextEditor from './TextEditor';
+import LegacyTextEditor from './LegacyTextEditor';
 import ClickOutside from './ClickOutside';
 import { InteractionMasksProps, InteractionMasksState } from '../../masks/InteractionMasks';
 
 type SharedInteractionMasksProps<R, K extends keyof R> = Pick<InteractionMasksProps<R, K>, 'scrollLeft' | 'scrollTop'>;
 type SharedInteractionMasksState = Pick<InteractionMasksState, 'firstEditorKeyPress'>;
 
-type ValueType<R> = R[keyof R];
-
-export interface Props<R, K extends keyof R> extends SharedInteractionMasksProps<R, K>, SharedInteractionMasksState, Omit<Dimension, 'zIndex'> {
+export interface EditorContainerProps<R, RK extends keyof R, CK extends keyof R> extends SharedInteractionMasksProps<R, RK>, SharedInteractionMasksState, Omit<Dimension, 'zIndex'> {
   rowIdx: number;
   rowData: R;
-  value: ValueType<R>;
-  column: CalculatedColumn<R>;
-  onGridKeyDown?(e: KeyboardEvent): void;
-  onCommit(e: CommitEvent<R>): void;
+  value: R[CK];
+  column: CalculatedColumn<R, unknown, CK>;
+  onCommit(e: CommitEvent<R, { [k: string]: R[CK] }>): void;
   onCommitCancel(): void;
 }
 
-interface State {
-  isInvalid: boolean;
-}
-
-export default class EditorContainer<R, K extends keyof R> extends React.Component<Props<R, K>, State> {
-  static displayName = 'EditorContainer';
+function getInitialValue<R, CK extends keyof R>({ key, value }: { key: string | null; value: R[CK] }): R[CK] | string {
+  // TODO: Do we need to activate editor when Delete, Backspace or other keys are pressed?
+  // These keys only make sense for a textbox editor. Activating a complex editor like a date-picker or a modal would not be intuitive
+  if (key === 'Delete' || key === 'Backspace') {
+    return '';
+  }
+  if (key === 'Enter') {
+    return value;
+  }
 
-  changeCommitted = false;
-  changeCanceled = false;
+  return key || value;
+}
 
-  private readonly editor = React.createRef<Editor>();
-  readonly state: Readonly<State> = { isInvalid: false };
+export default function EditorContainer<R, RK extends keyof R, CK extends keyof R = keyof R>({
+  rowIdx,
+  column,
+  rowData,
+  width,
+  height,
+  left,
+  top,
+  onCommit,
+  onCommitCancel,
+  scrollLeft,
+  scrollTop,
+  firstEditorKeyPress,
+  ...props
+}: EditorContainerProps<R, RK, CK>) {
+  const editorRef = useRef<Editor<{ [k: string]: R[CK] }>>(null);
+  const inputRef = useRef<HTMLInputElement>(null);
+  const [isInvalid, setIsInvalid] = useState(false);
+  const [value, setValue] = useState(() => getInitialValue({ key: firstEditorKeyPress, value: props.value }));
+  const prevScrollLeft = useRef(scrollLeft);
+  const prevScrollTop = useRef(scrollTop);
+  const { enableNewEditor, getRowMetaData, editor } = column;
+
+  const getInputNode = useCallback(() => {
+    return enableNewEditor
+      ? inputRef.current
+      : editorRef.current && editorRef.current.getInputNode();
+  }, [enableNewEditor]);
+
+  useEffect(() => {
+    const inputNode = getInputNode();
```

Comment:
My only concern is that if we put this `useEffect` in the editor container, the RDG might delegate too much for the `CustomEditor`. We should leave it to the consumer project to decide if they want to focus it by entering the editor or not. Just to be flexible. That's the reason why I added that hook.

We could consider passing the `firstEditorKeyPress` down to the `CustomEditor` as well to give the `CustomEditor` more flexibility. As currently for the line#73, it has a bug that if you input the value like `1` and `2` and `3`. the result would be just `23`. as the input `1` will open the editor and make the input selecting all, and the second input `2` will replace the `1` and set the cursor to the end which results in having `23` in the input.

Reply:
> My only concern is that if we put this useEffect in the editor container, the RDG might delegate too much for the CustomEditor. We should leave it to the consumer project to decide if they want to focus it by entering the editor or not. Just to be flexible

This is totally optional with the new editor api. If `editorRef(previously inputRef)` is set and it is an `input/select` element then we add additional behavior which is 
- prevent navigation based on the caret position
- focus/select element


> We could consider passing the firstEditorKeyPress down to the CustomEditor as well to give the CustomEditor more flexibility if we don't use the above hook that I mentioned. As currently for the line#73, it has a bug that if you input the value like 1 and 2 and 3. the result would be just 23. as the input 1 will open the editor and make the input selecting all, and the second input 2 will replace the 1 and set the cursor to the end which results in having 23 in the input.

I am not sure if we should open editor or every key. Activating a complex editor like a date-picker does not make much sense. May be we can somehow allow users to define the keys that should open the editor. I am not too sure at this point and we need some investigation. I like how slick grid editors work

https://mleibman.github.io/SlickGrid/examples/example3-editing.html

I think the new API is easy to work with. We only introduced two new props
- `onChange` - Call it when the value is changed. This way `getInputNode` is not required
- `editorRef` - set it if you want the grid to handle keyboard navigation for input elements and selection for `input/select` elements

This PR does not modify how editors are opened and this is still up for discussion

Thanks for playing with the idea of using hooks for navigation but it seems like it may be adding additional complexity. @MayhemYDG what do you think about the two approaches. Other ideas?

Reply:
```js
const columns = {
  ...column,
  onCellInput({ commit, ...data }) {
    // let users handle it here, per column
    commit(value);
    return true; // to open the editor
  }
}
```

I think this is what we should do. In most cases users will only use a text input column, so we can provide a simple default handler for that. `onCellInput: defaultCellInputHandler`.
If `onCellInput` is missing, then nothing will happen.

This handler could also handle backspace/delete, as well as pasting data.

explicit behavior > implicit assumptions, even if it means writing more code.

>it has a bug that if you input the value like 1 and 2 and 3. the result would be just 23. as the input 1 will open the editor and make the input selecting all, and the second input 2 will replace the 1 and set the cursor to the end which results in having 23 in the input.

Sounds like what we want is a `openReason: 'click' | 'input' | anything else?`. Then you can adjust the initial state of your editor as desired.